File: src/engine.cpp
#include "hanfe/engine.hpp"

#include <linux/input-event-codes.h>

#include <algorithm>
#include <cerrno>
#include <cstring>
#include <stdexcept>
#include <sys/ioctl.h>
#include <unistd.h>

#include "hanfe/util.hpp"

namespace hanfe {
namespace {

const std::vector<int> SHIFT_KEYS = {KEY_LEFTSHIFT, KEY_RIGHTSHIFT};
const std::vector<int> CTRL_KEYS = {KEY_LEFTCTRL, KEY_RIGHTCTRL};
const std::vector<int> ALT_KEYS = {KEY_LEFTALT, KEY_RIGHTALT};
const std::vector<int> META_KEYS = {KEY_LEFTMETA, KEY_RIGHTMETA};

std::vector<int> combine(const std::vector<int>& a, const std::vector<int>& b) {
    std::vector<int> result = a;
    result.insert(result.end(), b.begin(), b.end());
    return result;
}

const std::vector<int> MODIFIER_KEYS = combine(combine(combine(SHIFT_KEYS, CTRL_KEYS), ALT_KEYS), META_KEYS);
const std::vector<int> ALWAYS_FORWARD = combine(combine(CTRL_KEYS, ALT_KEYS), META_KEYS);

bool is_key_press(const input_event& event) { return event.value == 1 || event.value == 2; }

bool is_key_release(const input_event& event) { return event.value == 0; }

}  // namespace

HanfeEngine::HanfeEngine(int device_fd, Layout layout, ToggleConfig toggle,
                         std::unique_ptr<FallbackEmitter> emitter)
    : device_fd_(device_fd),
      layout_(std::move(layout)),
      toggle_(std::move(toggle)),
      emitter_(std::move(emitter)),
      mode_(toggle_.default_mode),
      preedit_text_("") {
    if (device_fd_ < 0) {
        throw std::runtime_error("Invalid device descriptor");
    }
    toggle_keys_.insert(toggle_.toggle_keys.begin(), toggle_.toggle_keys.end());
    for (int code : MODIFIER_KEYS) {
        modifier_state_[code] = false;
        forwarded_modifiers_[code] = false;
    }
}

void HanfeEngine::run() {
    if (ioctl(device_fd_, EVIOCGRAB, 1) < 0) {
        throw std::runtime_error("Failed to grab device: " + std::string(strerror(errno)));
    }
    try {
        input_event event{};
        while (true) {
            ssize_t n = ::read(device_fd_, &event, sizeof(event));
            if (n < 0) {
                if (errno == EAGAIN || errno == EINTR) {
                    continue;
                }
                throw std::runtime_error("Failed to read input event: " +
                                         std::string(strerror(errno)));
            }
            if (n == 0) {
                break;
            }
            if (n != sizeof(event)) {
                continue;
            }
            process_event(event);
        }
    } catch (...) {
        ioctl(device_fd_, EVIOCGRAB, 0);
        emitter_->close();
        throw;
    }
    ioctl(device_fd_, EVIOCGRAB, 0);
    emitter_->close();
}

void HanfeEngine::process_event(const input_event& event) {
    if (event.type != EV_KEY) {
        if (mode_ == InputMode::Latin) {
            emitter_->forward_event(event);
        }
        return;
    }

    int code = event.code;
    if (toggle_keys_.count(code) > 0) {
        if (event.value == 1) {
            toggle_mode();
        }
        return;
    }

    if (std::find(MODIFIER_KEYS.begin(), MODIFIER_KEYS.end(), code) != MODIFIER_KEYS.end()) {
        handle_modifier(event);
        return;
    }

    if (mode_ == InputMode::Latin) {
        forward_key_event(event);
        return;
    }

    if (code == KEY_BACKSPACE) {
        handle_backspace(event);
        return;
    }

    if (is_key_release(event)) {
        handle_key_release(event);
        return;
    }

    handle_key_press(event);
}

void HanfeEngine::handle_modifier(const input_event& event) {
    int code = event.code;
    bool press = is_key_press(event);
    bool release = is_key_release(event);

    if (press) {
        modifier_state_[code] = true;
    } else if (release) {
        modifier_state_[code] = false;
    }

    if (mode_ == InputMode::Latin ||
        std::find(ALWAYS_FORWARD.begin(), ALWAYS_FORWARD.end(), code) != ALWAYS_FORWARD.end()) {
        forward_key_event(event);
        forwarded_modifiers_[code] = press && !release;
        return;
    }

    if (release && forwarded_modifiers_[code]) {
        set_forwarded_modifier(code, false);
    }
}

void HanfeEngine::handle_backspace(const input_event& event) {
    if (mode_ == InputMode::Latin) {
        forward_key_event(event);
        return;
    }
    if (is_key_release(event)) {
        if (forwarded_keys_.count(KEY_BACKSPACE)) {
            forward_key_event(event);
        }
        return;
    }

    auto new_preedit = composer_.backspace();
    if (new_preedit) {
        replace_preedit(*new_preedit);
        return;
    }
    commit_preedit();
    forward_key_event(event);
}

void HanfeEngine::handle_key_release(const input_event& event) {
    int code = event.code;
    if (forwarded_keys_.count(code) > 0) {
        forward_key_event(event);
    }
}

void HanfeEngine::handle_key_press(const input_event& event) {
    int code = event.code;
    if (modifiers_active(ALWAYS_FORWARD)) {
        commit_preedit();
        ensure_shift_forwarded();
        forward_key_event(event);
        return;
    }

    const LayoutSymbol* symbol = layout_.translate(code, shift_active());
    if (!symbol) {
        commit_preedit();
        ensure_shift_forwarded();
        forward_key_event(event);
        return;
    }

    switch (symbol->kind) {
        case SymbolKind::Passthrough: {
            if (symbol->commit_before) {
                commit_preedit();
            }
            ensure_shift_forwarded();
            forward_key_event(event);
            return;
        }
        case SymbolKind::Text: {
            if (symbol->commit_before) {
                commit_preedit();
            }
            send_text(symbol->text);
            return;
        }
        case SymbolKind::Jamo: {
            CompositionResult result = composer_.feed(symbol->jamo, symbol->role);
            if (!result.commit.empty()) {
                commit_text(result.commit);
            }
            if (result.preedit != preedit_text_) {
                replace_preedit(result.preedit);
            }
            return;
        }
    }
}

void HanfeEngine::forward_key_event(const input_event& event) {
    emitter_->forward_event(event);
    if (is_key_press(event)) {
        forwarded_keys_.insert(event.code);
    } else if (is_key_release(event)) {
        forwarded_keys_.erase(event.code);
    }
}

bool HanfeEngine::modifiers_active(const std::vector<int>& subset) const {
    for (int code : subset) {
        auto it = modifier_state_.find(code);
        if (it != modifier_state_.end() && it->second) {
            return true;
        }
    }
    return false;
}

bool HanfeEngine::shift_active() const { return modifiers_active(SHIFT_KEYS); }

void HanfeEngine::ensure_shift_forwarded() {
    for (int code : SHIFT_KEYS) {
        if (modifier_state_.at(code) && !forwarded_modifiers_.at(code)) {
            set_forwarded_modifier(code, true);
        }
    }
}

void HanfeEngine::set_forwarded_modifier(int code, bool pressed) {
    auto it = forwarded_modifiers_.find(code);
    if (it != forwarded_modifiers_.end() && it->second == pressed) {
        return;
    }
    emitter_->send_key_state(code, pressed);
    forwarded_modifiers_[code] = pressed;
}

std::vector<int> HanfeEngine::suspend_forwarded_modifiers() {
    std::vector<int> suspended;
    for (auto& [code, forwarded] : forwarded_modifiers_) {
        if (forwarded) {
            set_forwarded_modifier(code, false);
            suspended.push_back(code);
        }
    }
    return suspended;
}

void HanfeEngine::restore_forwarded_modifiers(const std::vector<int>& codes) {
    for (int code : codes) {
        if (modifier_state_.at(code)) {
            set_forwarded_modifier(code, true);
        }
    }
}

void HanfeEngine::toggle_mode() {
    commit_preedit();
    mode_ = (mode_ == InputMode::Hangul) ? InputMode::Latin : InputMode::Hangul;
}

void HanfeEngine::commit_text(const std::string& text) {
    if (text.empty()) {
        return;
    }
    replace_preedit("");
    send_text(text);
}

void HanfeEngine::commit_preedit() {
    std::string commit = composer_.flush();
    if (commit.empty() && preedit_text_.empty()) {
        return;
    }
    replace_preedit("");
    if (!commit.empty()) {
        send_text(commit);
    }
}

void HanfeEngine::replace_preedit(const std::string& new_text) {
    if (new_text == preedit_text_) {
        return;
    }
    auto suspended = suspend_forwarded_modifiers();
    if (!preedit_text_.empty()) {
        size_t old_count = utf8_to_u32(preedit_text_).size();
        if (old_count > 0) {
            emitter_->send_backspace(static_cast<int>(old_count));
        }
    }
    if (!new_text.empty()) {
        emitter_->send_text(new_text);
    }
    preedit_text_ = new_text;
    restore_forwarded_modifiers(suspended);
}

void HanfeEngine::send_text(const std::string& text) {
    if (text.empty()) {
        return;
    }
    auto suspended = suspend_forwarded_modifiers();
    emitter_->send_text(text);
    restore_forwarded_modifiers(suspended);
}

}  // namespace hanfe

File: internal/engine/engine.go
package engine

import (
	"fmt"
	"strings"
	"syscall"

	"github.com/gg582/hanfe/internal/backend"
	"github.com/gg582/hanfe/internal/config"
	"github.com/gg582/hanfe/internal/emitter"
	"github.com/gg582/hanfe/internal/hangul"
	"github.com/gg582/hanfe/internal/layout"
	"github.com/gg582/hanfe/internal/linux"
	"github.com/gg582/hanfe/internal/types"
	"github.com/gg582/hanfe/internal/util"
	"golang.org/x/sys/unix"
)

type ModeSpec struct {
	Name     string
	Kind     types.InputMode
	Layout   *layout.Layout
	Database backend.Database
}

type Engine struct {
	deviceFD           int
	modes              []ModeSpec
	modeIndex          int
	toggleChords       []config.ToggleChord
	emitter            emitter.Output
	hangulComposers    map[int]*hangul.HangulComposer
	modifierState      map[uint16]bool
	forwardedModifiers map[uint16]bool
	forwardedKeys      map[uint16]struct{}
	preedit            string
	pinyinBuffer       string
}

var (
	shiftKeys = []uint16{uint16(linux.KeyLeftShift), uint16(linux.KeyRightShift)}
	ctrlKeys  = []uint16{uint16(linux.KeyLeftCtrl), uint16(linux.KeyRightCtrl)}
	altKeys   = []uint16{uint16(linux.KeyLeftAlt), uint16(linux.KeyRightAlt)}
	metaKeys  = []uint16{uint16(linux.KeyLeftMeta), uint16(linux.KeyRightMeta)}
)

func combine(a, b []uint16) []uint16 {
	result := make([]uint16, 0, len(a)+len(b))
	result = append(result, a...)
	result = append(result, b...)
	return result
}

var (
	modifierKeys  = combine(combine(combine(shiftKeys, ctrlKeys), altKeys), metaKeys)
	alwaysForward = combine(combine(ctrlKeys, altKeys), metaKeys)
)

func NewEngine(deviceFD int, modes []ModeSpec, toggle config.ToggleConfig, emitter emitter.Output) (*Engine, error) {
	if len(modes) == 0 {
		return nil, fmt.Errorf("no input modes configured")
	}

	eng := &Engine{
		deviceFD:           deviceFD,
		modes:              modes,
		toggleChords:       toggle.Chords,
		emitter:            emitter,
		hangulComposers:    make(map[int]*hangul.HangulComposer),
		modifierState:      make(map[uint16]bool),
		forwardedModifiers: make(map[uint16]bool),
		forwardedKeys:      make(map[uint16]struct{}),
	}
	for _, code := range modifierKeys {
		eng.modifierState[code] = false
		eng.forwardedModifiers[code] = false
	}
	for _, chord := range toggle.Chords {
		for _, group := range chord.ModifierGroups {
			for _, code := range group {
				if _, ok := eng.modifierState[code]; !ok {
					eng.modifierState[code] = false
					eng.forwardedModifiers[code] = false
				}
			}
		}
	}
	defaultIndex := 0
	if toggle.DefaultMode != "" {
		for idx, mode := range modes {
			if strings.EqualFold(mode.Name, toggle.DefaultMode) {
				defaultIndex = idx
				break
			}
		}
	}
	eng.modeIndex = defaultIndex

	for idx, mode := range modes {
		if mode.Kind == types.ModeHangul {
			eng.hangulComposers[idx] = hangul.NewHangulComposer()
		}
	}
	return eng, nil
}

func (e *Engine) Run() error {
	if err := linux.IoctlSetInt(e.deviceFD, linux.EVIOCGRAB, 1); err != nil {
		return fmt.Errorf("grab device: %w", err)
	}
	defer linux.IoctlSetInt(e.deviceFD, linux.EVIOCGRAB, 0)
	defer e.emitter.Close()

	size := util.InputEventSize()
	pollFDs := []unix.PollFd{{Fd: int32(e.deviceFD), Events: unix.POLLIN}}
	for {
		var ev util.InputEvent
		buf := ev.Bytes()
		n, err := syscall.Read(e.deviceFD, buf)
		if err != nil {
			if err == syscall.EINTR {
				continue
			}
			if err == syscall.EAGAIN {
				if pollErr := waitForReadable(pollFDs); pollErr != nil {
					return fmt.Errorf("poll input device: %w", pollErr)
				}
				continue
			}
			return fmt.Errorf("read input event: %w", err)
		}
		if n == 0 {
			return nil
		}
		if n != size {
			continue
		}
		if err := e.processEvent(&ev); err != nil {
			return err
		}
	}
}

func waitForReadable(pollFDs []unix.PollFd) error {
	for {
		for i := range pollFDs {
			pollFDs[i].Revents = 0
		}
		n, err := unix.Poll(pollFDs, -1)
		if err == syscall.EINTR {
			continue
		}
		if err != nil {
			return err
		}
		if n == 0 {
			continue
		}
		ready := false
		for _, fd := range pollFDs {
			if fd.Revents&(unix.POLLERR|unix.POLLHUP|unix.POLLNVAL) != 0 {
				return syscall.EIO
			}
			if fd.Revents&(unix.POLLIN|unix.POLLPRI) != 0 {
				ready = true
			}
		}
		if ready {
			return nil
		}
	}
}

func (e *Engine) processEvent(event *util.InputEvent) error {
	if event.Type != linux.EvKey {
		if e.currentModeKind() == types.ModeLatin {
			return e.forwardKeyEvent(event)
		}
		return nil
	}

	if e.shouldToggle(event) {
		if isKeyPress(event) {
			return e.toggleMode()
		}
		return nil
	}

	code := event.Code
	if contains(modifierKeys, code) {
		return e.handleModifier(event)
	}

	if e.currentModeKind() == types.ModeLatin {
		return e.forwardKeyEvent(event)
	}

	if code == uint16(linux.KeyBackspace) {
		return e.handleBackspace(event)
	}

	if isKeyRelease(event) {
		return e.handleKeyRelease(event)
	}

	return e.handleKeyPress(event)
}

func (e *Engine) handleModifier(event *util.InputEvent) error {
	code := event.Code
	press := isKeyPress(event)
	release := isKeyRelease(event)

	if press {
		e.modifierState[code] = true
	} else if release {
		e.modifierState[code] = false
	}

	if e.currentModeKind() == types.ModeLatin || contains(alwaysForward, code) {
		if err := e.forwardKeyEvent(event); err != nil {
			return err
		}
		e.forwardedModifiers[code] = press && !release
		return nil
	}

	if release && e.forwardedModifiers[code] {
		e.setForwardedModifier(code, false)
	}
	return nil
}

func (e *Engine) handleBackspace(event *util.InputEvent) error {
	switch e.currentModeKind() {
	case types.ModeLatin, types.ModeKana:
		return e.forwardKeyEvent(event)
	case types.ModeDatabase:
		return e.handleDatabaseBackspace(event)
	}
	if isKeyRelease(event) {
		if _, ok := e.forwardedKeys[event.Code]; ok {
			return e.forwardKeyEvent(event)
		}
		return nil
	}
	if composer := e.currentComposerIfHangul(); composer != nil {
		if newPreedit, ok := composer.Backspace(); ok {
			return e.replacePreedit(newPreedit)
		}
	}
	if err := e.commitPreedit(); err != nil {
		return err
	}
	return e.forwardKeyEvent(event)
}

func (e *Engine) handleKeyRelease(event *util.InputEvent) error {
	if _, ok := e.forwardedKeys[event.Code]; ok {
		return e.forwardKeyEvent(event)
	}
	return nil
}

func (e *Engine) handleKeyPress(event *util.InputEvent) error {
	mode := e.currentMode()
	if mode.Kind == types.ModeDatabase {
		return e.handleDatabaseKeyPress(event)
	}

	if e.modifiersActive(alwaysForward) {
		if err := e.commitPreedit(); err != nil {
			return err
		}
		if err := e.ensureShiftForwarded(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	}

	if mode.Layout == nil {
		if err := e.commitPreedit(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	}

	symbol := mode.Layout.Translate(event.Code, e.shiftActive())
	if symbol == nil {
		if err := e.commitPreedit(); err != nil {
			return err
		}
		if err := e.ensureShiftForwarded(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	}

	switch symbol.Kind {
	case layout.SymbolPassthrough:
		if symbol.CommitBefore {
			if err := e.commitPreedit(); err != nil {
				return err
			}
		}
		if err := e.ensureShiftForwarded(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	case layout.SymbolText:
		if symbol.CommitBefore {
			if err := e.commitPreedit(); err != nil {
				return err
			}
		}
		return e.sendText(symbol.Text)
	case layout.SymbolJamo:
		composer := e.currentComposer()
		result := composer.Feed(symbol.Jamo, symbol.Role)
		if result.Commit != "" {
			if err := e.commitText(result.Commit); err != nil {
				return err
			}
		}
		if result.Preedit != e.preedit {
			if err := e.replacePreedit(result.Preedit); err != nil {
				return err
			}
		}
		return nil
	default:
		return nil
	}
}

func (e *Engine) forwardKeyEvent(event *util.InputEvent) error {
	if err := e.emitter.ForwardEvent(event); err != nil {
		return err
	}
	if isKeyPress(event) {
		e.forwardedKeys[event.Code] = struct{}{}
	} else if isKeyRelease(event) {
		delete(e.forwardedKeys, event.Code)
	}
	return nil
}

func (e *Engine) modifiersActive(codes []uint16) bool {
	for _, code := range codes {
		if e.modifierState[code] {
			return true
		}
	}
	return false
}

func (e *Engine) shiftActive() bool {
	return e.modifiersActive(shiftKeys)
}

func (e *Engine) modifierGroupsActive(groups [][]uint16) bool {
	for _, group := range groups {
		active := false
		for _, code := range group {
			if e.modifierState[code] {
				active = true
				break
			}
		}
		if !active {
			return false
		}
	}
	return true
}

func (e *Engine) shouldToggle(event *util.InputEvent) bool {
	if !isKeyPress(event) {
		return false
	}
	code := event.Code
	for _, chord := range e.toggleChords {
		if chord.Key != code {
			continue
		}
		if len(chord.ModifierGroups) == 0 || e.modifierGroupsActive(chord.ModifierGroups) {
			return true
		}
	}
	return false
}

func (e *Engine) currentMode() ModeSpec {
	return e.modes[e.modeIndex]
}

func (e *Engine) currentModeKind() types.InputMode {
	return e.modes[e.modeIndex].Kind
}

func (e *Engine) currentComposer() *hangul.HangulComposer {
	composer, ok := e.hangulComposers[e.modeIndex]
	if !ok || composer == nil {
		composer = hangul.NewHangulComposer()
		e.hangulComposers[e.modeIndex] = composer
	}
	return composer
}

func (e *Engine) currentComposerIfHangul() *hangul.HangulComposer {
	if e.currentModeKind() != types.ModeHangul {
		return nil
	}
	return e.currentComposer()
}

func (e *Engine) ensureShiftForwarded() error {
	for _, code := range shiftKeys {
		if e.modifierState[code] && !e.forwardedModifiers[code] {
			if err := e.setForwardedModifier(code, true); err != nil {
				return err
			}
		}
	}
	return nil
}

func (e *Engine) setForwardedModifier(code uint16, pressed bool) error {
	if current := e.forwardedModifiers[code]; current == pressed {
		return nil
	}
	if err := e.emitter.SendKeyState(code, pressed); err != nil {
		return err
	}
	e.forwardedModifiers[code] = pressed
	return nil
}

func (e *Engine) suspendForwardedModifiers() ([]uint16, error) {
	var suspended []uint16
	for code, forwarded := range e.forwardedModifiers {
		if forwarded {
			if err := e.setForwardedModifier(code, false); err != nil {
				return suspended, err
			}
			suspended = append(suspended, code)
		}
	}
	return suspended, nil
}

func (e *Engine) restoreForwardedModifiers(codes []uint16) {
	for _, code := range codes {
		if e.modifierState[code] {
			_ = e.setForwardedModifier(code, true)
		}
	}
}

func (e *Engine) toggleMode() error {
	if err := e.commitPreedit(); err != nil {
		return err
	}
	if len(e.modes) == 0 {
		return nil
	}
	e.modeIndex = (e.modeIndex + 1) % len(e.modes)
	e.pinyinBuffer = ""
	return e.replacePreedit("")
}

func (e *Engine) commitText(text string) error {
	if text == "" {
		return nil
	}
	if err := e.replacePreedit(""); err != nil {
		return err
	}
	return e.sendText(text)
}

func (e *Engine) commitPreedit() error {
	mode := e.currentMode()
	switch mode.Kind {
	case types.ModeHangul:
		composer := e.currentComposer()
		commit := composer.Flush()
		if commit == "" && e.preedit == "" {
			return nil
		}
		if err := e.replacePreedit(""); err != nil {
			return err
		}
		if commit != "" {
			if err := e.sendText(commit); err != nil {
				return err
			}
		}
		return nil
	case types.ModeDatabase:
		return e.commitPinyinBuffer()
	default:
		if e.preedit == "" {
			return nil
		}
		return e.replacePreedit("")
	}
}

func (e *Engine) handleDatabaseKeyPress(event *util.InputEvent) error {
	if e.modifiersActive(alwaysForward) {
		if err := e.commitPreedit(); err != nil {
			return err
		}
		if err := e.ensureShiftForwarded(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	}

	code := event.Code
	if code >= uint16(linux.KeyA) && code <= uint16(linux.KeyZ) {
		ch := rune('a' + int(code-uint16(linux.KeyA)))
		e.pinyinBuffer += string(ch)
		return e.replacePreedit(e.pinyinBuffer)
	}

	if code >= uint16(linux.Key1) && code <= uint16(linux.Key5) {
		ch := rune('1' + int(code-uint16(linux.Key1)))
		e.pinyinBuffer += string(ch)
		return e.replacePreedit(e.pinyinBuffer)
	}

	if code == uint16(linux.KeyApostrophe) {
		e.pinyinBuffer += "'"
		return e.replacePreedit(e.pinyinBuffer)
	}

	switch code {
	case uint16(linux.KeySpace), uint16(linux.KeyEnter), uint16(linux.KeyTab):
		if err := e.commitPinyinBuffer(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	default:
		if err := e.commitPinyinBuffer(); err != nil {
			return err
		}
		if err := e.ensureShiftForwarded(); err != nil {
			return err
		}
		return e.forwardKeyEvent(event)
	}
}

func (e *Engine) handleDatabaseBackspace(event *util.InputEvent) error {
	if isKeyRelease(event) {
		return nil
	}
	if e.pinyinBuffer == "" {
		return e.forwardKeyEvent(event)
	}
	runes := []rune(e.pinyinBuffer)
	if len(runes) == 0 {
		return e.replacePreedit("")
	}
	e.pinyinBuffer = string(runes[:len(runes)-1])
	return e.replacePreedit(e.pinyinBuffer)
}

func (e *Engine) commitPinyinBuffer() error {
	if e.pinyinBuffer == "" {
		if e.preedit != "" {
			return e.replacePreedit("")
		}
		return nil
	}
	mode := e.currentMode()
	text := e.pinyinBuffer
	if mode.Database.Available() {
		if candidate, ok := mode.Database.Lookup(text); ok {
			text = candidate
		}
	}
	if err := e.replacePreedit(""); err != nil {
		return err
	}
	if err := e.sendText(text); err != nil {
		return err
	}
	e.pinyinBuffer = ""
	return nil
}

func (e *Engine) replacePreedit(newText string) error {
	if newText == e.preedit {
		return nil
	}
	if !e.emitter.SupportsPreedit() {
		e.preedit = newText
		return nil
	}
	suspended, err := e.suspendForwardedModifiers()
	if err != nil {
		return err
	}
	if e.preedit != "" {
		count := countRunes(e.preedit)
		if count > 0 {
			if err := e.emitter.SendBackspace(count); err != nil {
				e.restoreForwardedModifiers(suspended)
				return err
			}
		}
	}
	if newText != "" {
		if err := e.emitter.SendText(newText); err != nil {
			e.restoreForwardedModifiers(suspended)
			return err
		}
	}
	e.preedit = newText
	e.restoreForwardedModifiers(suspended)
	return nil
}

func (e *Engine) sendText(text string) error {
	if text == "" {
		return nil
	}
	suspended, err := e.suspendForwardedModifiers()
	if err != nil {
		return err
	}
	err = e.emitter.SendText(text)
	e.restoreForwardedModifiers(suspended)
	return err
}

func countRunes(s string) int {
	count := 0
	for range s {
		count++
	}
	return count
}

func isKeyPress(ev *util.InputEvent) bool {
	return ev.Value == 1 || ev.Value == 2
}

func isKeyRelease(ev *util.InputEvent) bool {
	return ev.Value == 0
}

func contains(list []uint16, code uint16) bool {
	for _, c := range list {
		if c == code {
			return true
		}
	}
	return false
}
